File: Excel/431 Challenge.R
library(tidyverse)
library(readxl)

input = read_excel("Excel/431 Top 3 Rankings.xlsx", range = "A1:H20")
test  = read_excel("Excel/431 Top 3 Rankings.xlsx", range = "J2:K5")

result = input %>%
  pivot_longer(cols = -c(1), names_to = "year", values_to = "result") %>%
  mutate(Rank = dense_rank(desc(result)) %>% as.numeric(), .by = year) %>%
  filter(Rank <= 3) %>%
  summarise(n = n_distinct(year), .by = c("Region", "Rank")) %>%
  mutate(check = n == max(n), .by = "Rank") %>%
  filter(check) %>%
  summarise(Regions = paste(Region, collapse = ", "), .by = "Rank") %>%
  arrange(Rank) 

identical(result, test)
# [1] TRUE

File: Excel/431 Challenge.py
import pandas as pd

input = pd.read_excel("431 Top 3 Rankings.xlsx", usecols="A:H", nrows=20)
test = pd.read_excel("431 Top 3 Rankings.xlsx",  usecols="J:K", nrows=3, skiprows=1)

result = input.melt(id_vars=["Region"], var_name="year", value_name="result")
result["Rank"] = result.groupby("year")["result"].rank("dense", ascending=False).astype('int64')
result = result[result["Rank"] <= 3]
result = result.groupby(["Region", "Rank"]).size().reset_index(name="n")
result = result.groupby("Rank").apply(lambda x: x[x["n"] == x["n"].max()]).reset_index(drop=True)
result["RegionNo"] = result["Region"].str.extract(r"(\d+)").astype(int)
result = result.sort_values(["Rank", "RegionNo"])

result = result.groupby("Rank")["Region"].apply(lambda x: ", ".join(x)).reset_index(name="Regions")

print(result.equals(test))